### DISCUSSION TITLE
How to unit-test jq scripts?

### DISCUSSION THREAD

**Discussion by techhazard** [jqlang/jq#3147] [Q&A]
I'm writing several jq scripts, and I would like to add tests for those. 

I have found the `--run-tests` argument that seems very useful, but it does not allow including other files (it assumes that's another test).

Here's what I expect to work
```jq
# myscript.jq
def rename_name_to_fullname:
  .fullname = .name | del(.name)
;
```

```jq
# myscript.test.jq
include "myscript" # <- this does not work, because it is interpreted as the program line of a test

# first test
rename_name_to_fullname  # program line
{ "name": "Jason Query"}    # input line
{ "fullname": "Jason Query"} # output line
```

Maybe the syntax for the `include` needs to be different in the test scripts, but I find the documentation of this testing setup to be very sparse in general.
Or is there another way to properly add tests?

**Comment by wellkilo** [comment]
The test file grammar expects the first non-comment line of each case to be the program, so a standalone include line is indeed treated as that test's program. Put the include/import on the same program line as the function call, and pass the module directory with -L:

    # myscript.test
    include "myscript"; rename_name_to_fullname
    {"name":"Jason Query"}
    {"fullname":"Jason Query"}

Run it as:

    jq -L . --run-tests myscript.test

Keep the reusable file named myscript.jq in the directory supplied to -L. You can also use a namespace:

    import "myscript" as m; m::rename_name_to_fullname
    {"name":"Jason Query"}
    {"fullname":"Jason Query"}

The important CLI detail is that --run-tests must be the last option, so -L must come before it. jq's own test drivers use the same ordering, for example jq -L "$mods" --run-tests optional.test: [tests/optionaltest](https://github.com/jqlang/jq/blob/master/tests/optionaltest). The manual documents both the test-file grammar and the -L module search path: [--run-tests](https://github.com/jqlang/jq/blob/master/docs/content/manual/v1.8/manual.yml#L316-L324).

For larger suites, a normal shell/pytest harness invoking jq -L . -f myscript.jq per fixture is often easier to structure, but it is not required for includes.
